### ISSUE TITLE
Customized fluentd image (fluentd v1.0 + fluent-plugin-elasticsearch -v 2.4.0) doesn't work

### ISSUE THREAD

**Issue by ahrtr** [fluent/fluentd#1837]
I built fluentd image using the following Dockerfile, 

```
FROM fluent/fluentd:v1.0-debian
USER root
WORKDIR /home/fluent
ENV PATH /home/fluent/.gem/ruby/2.3.0/bin:$PATH
RUN buildDeps="sudo make gcc g++ libc-dev ruby-dev libffi-dev" \
     && apt-get update \
     && apt-get upgrade -y \
     && apt-get install \
     -y --no-install-recommends \
     $buildDeps \
    && echo 'gem: --no-document' >> /etc/gemrc \
    && gem install fluent-plugin-secure-forward \
    && gem install fluent-plugin-record-reformer \
    && gem install fluent-plugin-elasticsearch -v 2.4.0 \
    && gem install fluent-plugin-kubernetes_metadata_filter \
    && gem install ffi \
    && gem install fluent-plugin-systemd -v 0.0.8 \
    && SUDO_FORCE_REMOVE=yes \
    apt-get purge -y --auto-remove \
                  -o APT::AutoRemove::RecommendsImportant=false \
                  $buildDeps \
 && rm -rf /var/lib/apt/lists/* \
    && gem sources --clear-all \
    && rm -rf /tmp/* /var/tmp/* /usr/lib/ruby/gems/*/cache/*.gem
# Copy configuration files
COPY ./conf/fluent.conf /fluentd/etc/
COPY ./conf/systemd.conf /fluentd/etc/
COPY ./conf/kubernetes.conf /fluentd/etc/
# Copy plugins
COPY plugins /fluentd/plugins/
COPY entrypoint.sh /fluentd/entrypoint.sh
# Environment variables
ENV FLUENTD_OPT=""
ENV FLUENTD_CONF="fluent.conf"
# jemalloc is memory optimization only available for td-agent
# td-agent is provided and QA'ed by treasuredata as rpm/deb/.. package
# -> td-agent (stable) vs fluentd (edge)
#ENV LD_PRELOAD="/usr/lib/libjemalloc.so.2"
# Run Fluentd
CMD ["/fluentd/entrypoint.sh"]
```

And then the configuration file for fluentd is as below,
```
      tag fluentd
      <parse>
        @type json
        time_key time
        keep_time_key true
      </parse>
      refresh_interval 5
    </source>
    <filter **>
        @type grep
        <regexp>
          key log
          pattern TESTDATA
        </regexp>
    </filter>
    <filter **>
        @type parser
        format json
        key_name log
        reserve_data true
        hash_value_field log
    </filter>
    <match fluentd>
      @type elasticsearch
      host elasticsearch
      port 9200
      flush_interval 10s
      index_name test-${.log.tenantid}-%Y%m%d
      <buffer tag, .log.tenantid, time>
         @type memory
         timekey 3600
      </buffer>
    </match>
```

Everything is fine, and I do not see any error logs, but the issue is that the fluentd can't receive any log data. 

Previously I built a fluentd image using fluentd:v0.12.33-debian and fluent-plugin-elasticsearch without specifying the version, and it worked. 

This time, I made the following two changes, but it doesn't work any more,
1. I replaced the fluentd base image with v1.0 and specified a version v2.4.0 for fluent-plugin-elasticsearch;
2. I replaced the ${log.tenantid} with ${.log.tenantid} per cosmo0920's suggestion.

**Comment by repeatedly** [comment]
> .log.tenantid

typo of `$.log.tenantid`?

**Comment by ahrtr** [comment]
Thanks for the response. I updated the configuration as below, but it did not work either,

```
    <match fluentd>
      @type elasticsearch
      host elasticsearch
      port 9200
      flush_interval 10s
      index_name test-${$.log.tenantid}-%Y%m%d
      <buffer tag, $.log.tenantid, time>
         @type memory
         timekey 3600
      </buffer>
    </match>
```

**Comment by ahrtr** [comment]
I tried fluentd v1.1.0 + fluentd-plugin-elasticsearch v2.5.0, and it did not work either. 

But it worked with fluentd v0.12.33 + fluentd-plugin-elasticsearch (without specifying the version).

**Comment by repeatedly** [comment]
> But it worked with fluentd v0.12.33 + fluentd-plugin-elasticsearch (without specifying the version).

It seems to be same with fluentd v1.1.0 + fluentd-plugin-elasticsearch v2.5.0 because if you don't specify version, gem install latest version.

**Comment by ahrtr** [comment]
In summary, I tested the following cases, and confirmed that only the first case works with target_index_key.  All other cases do not work with either target_indx_key or $.log.tenantid. Have you ever tested fluentd-plugin-elasticsearch 2.5 + fluentd 1.x?

1. fluentd v0.12.33 + fluentd-plugin-elasticsearch v2.5.0
2. fluentd v1.0 +  fluentd-plugin-elasticsearch v2.5.0
3. fluentd v1.1.0 +  fluentd-plugin-elasticsearch v2.5.0
4. fluentd v1.0 +  fluentd-plugin-elasticsearch v2.4.0

**Comment by ahrtr** [comment]
For your information, when I started the fluentd (fluentd v0.12.33 + fluentd-plugin-elasticsearch v2.5.0), I saw lots of the following log messages on startup, but I did not similar log messages when starting fluentd of case 2 ~ 4. 

```
2018-01-30 12:32:07 +0000 [info]: #0 following tail of /var/lib/docker/containers/ccac8bde16d0145dba9c2ba65ec19f458db602b13f371a543262b95182e10e3d/ccac8bde16d0145dba9c2ba65ec19f458db602b13f371a543262b95182e10e3d-json.log
2018-01-30 12:32:07 +0000 [info]: #0 following tail of /var/lib/docker/containers/ba773e8c630c0d51898e87775c04852c13bc8361088a8b4d4aa9786f4b51e108/ba773e8c630c0d51898e87775c04852c13bc8361088a8b4d4aa9786f4b51e108-json.log
2018-01-30 12:32:07 +0000 [info]: #0 following tail of /var/lib/docker/containers/f52353efde8f1e23d00d84df65dae07b6d830ce40af6c94fd940b197f1488628/f52353efde8f1e23d00d84df65dae07b6d830ce40af6c94fd940b197f1488628-json.log

```

**Comment by repeatedly** [comment]
>  Have you ever tested fluentd-plugin-elasticsearch 2.5 + fluentd 1.x?

I didn't test because I'm not elasticsearch user but some users use fluentd v1.x with latest elaticsearch plugin.

> fluentd v0.12.33 + fluentd-plugin-elasticsearch v2.5.0

This doesn't work because fluent-plugin-elasticsearch requires fluentd v0.14.20 or later.

https://rubygems.org/gems/fluent-plugin-elasticsearch/versions/2.5.0

**Comment by ahrtr** [comment]
Thanks for the info. 

Based on my test, fluentd v0.12.33 works with fluentd-plugin-elasticsearch v2.5.0. Please see the following log messages.  Anway, it isn't the point, any suggestions on how to debug the issue of fluentd 1.x + fluentd-plugin-elasticsearch v2.5.0?
```

2018-01-30 12:32:06 +0000 [info]: starting fluentd-1.1.0 pid=1 ruby="2.3.3"
2018-01-30 12:32:06 +0000 [info]: spawn command to main:  cmdline=["/usr/bin/ruby2.3", "-Eascii-8bit:ascii-8bit", "/usr/local/bin/fluentd", "-c", "/fluentd/etc/fluentd.conf", "-p", "/fluentd/plugins", "--under-supervisor"]
2018-01-30 12:32:07 +0000 [info]: gem 'fluent-plugin-elasticsearch' version '2.5.0'
2018-01-30 12:32:07 +0000 [info]: gem 'fluent-plugin-kubernetes_metadata_filter' version '1.0.1'
2018-01-30 12:32:07 +0000 [info]: gem 'fluent-plugin-record-reformer' version '0.9.1'
2018-01-30 12:32:07 +0000 [info]: gem 'fluent-plugin-secure-forward' version '0.4.5'
2018-01-30 12:32:07 +0000 [info]: gem 'fluent-plugin-systemd' version '0.0.8'
2018-01-30 12:32:07 +0000 [info]: gem 'fluentd' version '1.1.0'
2018-01-30 12:32:07 +0000 [info]: gem 'fluentd' version '0.12.33'
2018-01-30 12:32:07 +0000 [info]: adding filter pattern="**" type="grep"
2018-01-30 12:32:07 +0000 [info]: adding filter pattern="**" type="parser"
2018-01-30 12:32:07 +0000 [info]: adding match pattern="fluentd" type="elasticsearch"
2018-01-30 12:32:07 +0000 [info]: #0 'flush_interval' is configured at out side of <buffer>. 'flush_mode' is set to 'interval' to keep existing behaviour
2018-01-30 12:32:07 +0000 [info]: adding source type="tail"
2018-01-30 12:32:07 +0000 [warn]: #0 'pos_file PATH' parameter is not set to a 'tail' source.
2018-01-30 12:32:07 +0000 [warn]: #0 this parameter is highly recommended to save the position to resume tailing.
2018-01-30 12:32:07 +0000 [info]: #0 starting fluentd worker pid=13 ppid=1 worker=0
2018-01-30 12:32:07 +0000 [info]: #0 following tail of /var/lib/docker/containers/ccac8bde16d0145dba9c2ba65ec19f458db602b13f371a543262b95182e10e3d/ccac8bde16d0145dba9c2ba65ec19f458db602b13f371a543262b95182e10e3d-json.log

```

**Comment by repeatedly** [comment]
> 2018-01-30 12:32:06 +0000 [info]: starting fluentd-1.1.0 pid=1 ruby="2.3.3"

Log say use fluentd 1.1.0, not 0.12.33

**Comment by ahrtr** [comment]
Yes, I saw the log entry as well. It's really confusing. I am pretty sure I built the image using fluentdv0.12.33 as the base image, and I saw fluentd v0.12.33 in the log as well. 
```

2018-01-30 12:32:06 +0000 [info]: starting fluentd-1.1.0 pid=1 ruby="2.3.3"
...
2018-01-30 12:32:07 +0000 [info]: gem 'fluentd' version '1.1.0'
2018-01-30 12:32:07 +0000 [info]: gem 'fluentd' version '0.12.33'
```

**Comment by repeatedly** [comment]
>  I am pretty sure I built the image using fluentdv0.12.33 as the base image

This point is described in the official article: https://docs.fluentd.org/v1.0/articles/plugin-management#plugin-version-management

> I saw fluentd v0.12.33 in the log as well.

Fluentd dumps all gem versions for debug purpose for users.
User can know expected version is installed or not. Generally, if you see multiple gem versions with same name, unexpected version is installed.

**Comment by ahrtr** [comment]
I did lots of tests, and eventually I confirmed that only "fluentd 0.12.33 + fluentd-plugin-elasticsearch v 2.5.0" works,  the Dockerfile is as below,

```
FROM fluent/fluentd:v0.12.33-debian

MAINTAINER Benjamin Wang <benjamin.w.wang@oracle.com>
USER root
WORKDIR /home/fluent
ENV PATH /home/fluent/.gem/ruby/2.3.0/bin:$PATH

RUN buildDeps="sudo make gcc g++ libc-dev ruby-dev libffi-dev" \
     && apt-get update \
     && apt-get upgrade -y \
     && apt-get install \
     -y --no-install-recommends \
     $buildDeps \
    && echo 'gem: --no-document' >> /etc/gemrc \
    && gem install fluent-plugin-elasticsearch -v 2.5.0 \
    && gem install fluent-plugin-systemd -v 0.0.8 \
    && SUDO_FORCE_REMOVE=yes \
    apt-get purge -y --auto-remove \
                  -o APT::AutoRemove::RecommendsImportant=false \
                  $buildDeps \
 && rm -rf /var/lib/apt/lists/* \
    && gem sources --clear-all \
    && rm -rf /tmp/* /var/tmp/* /usr/lib/ruby/gems/*/cache/*.gem

# Copy configuration files
COPY ./conf/fluent.conf /fluentd/etc/
COPY ./conf/systemd.conf /fluentd/etc/
COPY ./conf/kubernetes.conf /fluentd/etc/

# Copy plugins
COPY plugins /fluentd/plugins/
COPY entrypoint.sh /fluentd/entrypoint.sh

# Environment variables
ENV FLUENTD_OPT=""
ENV FLUENTD_CONF="fluent.conf"

# Run Fluentd
CMD ["/fluentd/entrypoint.sh"]
```

**Comment by ahrtr** [comment]
It's right based on the the following link provided by repeadedly yesterday,
`https://rubygems.org/gems/fluent-plugin-elasticsearch/versions/2.5.0 `

But the reality is that I built the fluentd image using fluent/fluentd:v0.12.33-debian + fluent-plugin-elasticsearch v 2.5.0 (You can see the Dockerfile in my previous post), and it works, although there were some strange log messages on startup. Based on the feedback from repeadedly, it means that unexpected version is installed. But anyhow, it works.  

> 2018-01-30 12:32:06 +0000 [info]: starting fluentd-1.1.0 pid=1 ruby="2.3.3"
> ...
> 2018-01-30 12:32:07 +0000 [info]: gem 'fluentd' version '1.1.0'
> 2018-01-30 12:32:07 +0000 [info]: gem 'fluentd' version '0.12.33'

When I change the base image to fluentd v1.1.0 and keep all other unchanged, there isn't any error message, and nor any strange messages any more, but it doesn't work. Fluentd can't read any log entries any longer.  

By the way, I tried fluentd v0.12.33 + fluentd-plugin-elasticsearch v1.13.0, and it did not work either. 

So I have to use fluentd v0.12.33 as the base image, and install fluent-plugin-elasticsearch v2.5.0 in the end.

**Comment by ahrtr** [comment]
Can anyone share a workable Dockerfile based on fluentd v1.x + fluentd-plugin-elasticsearch v2.x?

**Comment by dannyk81** [comment]
This works just fine:
```
FROM fluent/fluentd:v1.1.0-debian-onbuild
USER root

RUN ulimit -n 65536

RUN buildDeps="sudo make gcc g++ libc-dev ruby-dev" \
 && apt-get update \
 && apt-get install -y --no-install-recommends $buildDeps \
 && gem install \
        fluent-plugin-elasticsearch \
        fluent-plugin-record-reformer \
        fluent-plugin-systemd \
        fluent-plugin-kubernetes_metadata_filter \
        fluent-plugin-detect-exceptions \
        fluent-plugin-prometheus \
        fluent-plugin-record-modifier \
        fluent-plugin-forest \
        fluent-plugin-genhashvalue \
        fluent-plugin-elasticsearch-timestamp-check \
 && gem sources --clear-all \
 && SUDO_FORCE_REMOVE=yes \
    apt-get purge -y --auto-remove \
                  -o APT::AutoRemove::RecommendsImportant=false \
                  $buildDeps \
 && rm -rf /var/lib/apt/lists/* \
           /home/fluent/.gem/ruby/2.3.0/cache/*.gem \
 && mkdir /fluentd/etc/conf.d

COPY entrypoint.sh /bin/

RUN chmod +x /bin/entrypoint.sh
```

Note that I'm using `fluent/fluentd:v1.1.0-debian-onbuild` image as base.

**Comment by ahrtr** [comment]
@dannyk81 Thanks for the info.  I tried your dockerfile, and it did not work either. Actually I tried to build my fluentd image using fluent/fluentd:v1.1.0-debian-onbuild as the base image 2 days ago, and it did not work.

Just as I pointed out earlier, I did lots of tests, and eventually it only works when I use fluent/fluentd:v0.12.33-debian as the base image, and it also works when I use the officially fluentd image fluent/fluentd-kubernetes-daemonset:v0.12-debian-elasticsearch.
